Break big lock into some tiny locks for containerStart

WeiZhang555 opened the pull request:
Don't involve code waiting for blocking channel in locked critical
section because it has potential risk of hanging forever.
## What problem is this PR trying to fix?

Recently I'm doing some works on exec drivers, after some failures, I realize that the lock usage in exec driver has big problem, currently I want to mention `docker start`.

The current work flow of `postContainersStart` for `POST /containers/{name:.*}/start` is like this shows:

```
postContainersStart()-->daemon.ContainerStart()-->daemon.containerStart()
->container.Lock()-->**daemon.waitForStart()**-->container.Unlock()
```

work flow of `daemon.waitForStart`:

```
StartMonitor()--->monitor.wait()-->BLOCK: waiting for startSignal (read from channel)
```

When will `startSignal` be released?

```
monitor.Start()---> daemon.Run() --> execdriver(native).Run()--->callback function
-->release startSginal (close channel)
```

So the problem here is, the locked area (critical section) contains blocking code, which will lead to great risk of pending for a long time even forever,  most obvious result is `docker ps` command will hang forever (I've seen a lot of issues about docker ps hang, may not due to lock but possible).
## Is this really possible? I've never seen this happen!

With a carefully designed and implemented exec driver such as our `native` exec driver, I'm sure this problem may never happen, but in the future we may import more exec drivers, or simply a developer like me is interested in implementing a custom exec driver and leave it in `daemon/execdrivers/` package, it will save us a lot of time of taking care of the locks, lock is complicated always and need careful treat.

The way I found this problem is that when I forget to invoke `hook.Start` callback function which will release the startSignal in my custom exec driver, then container lock will be holden by the container forever, and hangs `docker ps`.
## How to solve this?

My solution is to break big lock into several tiny locks, do NOT lock the pending area, but lock the sub-functions. An extra benefit for smaller locks is that `docker ps` can have more chances to get its response.
## Other things

I was intending to break it into smaller locks than current patch, but it's a little complicated, and maybe some locks can be utterly removed but I'm also not sure. Maybe we can leave it as future work.

I've done some benchmarks on this, with 3 parallel processes iteratively start container, and no error observed. So I guess it's fine, maybe maintainers can look into this and give more suggestions.

Don't know who to ping, so ping @cpuguy83 @vdemeester , thank you!

Signed-off-by: Zhang Wei zhangwei555@huawei.com

WeiZhang555 committed:
Break big lock into some tiny locks

Don't involve code waiting for blocking channel in locked critical
section because it has potential risk of hanging forever.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

calavera commented:
it looks like this is not completely correct. The restarting tests are blocked because of this change. I'm not against it, but we should be very careful.

WeiZhang555 commented:
@calavera Yes, you're right. I'm still researching the root cause of blocked test case but got nothing, it's complicated and dangerous manipulating locks, I'll keep working on this and make sure all test cases passed.

WeiZhang555 commented:
Janky is not happy, but others are fine now.

thaJeztah commented:
Restarted Janky :+1:

calavera commented:
@WeiZhang555 can you test this patch with https://github.com/crosbymichael/docker-stress ? I'd be more confident on its behavior after seen docker-stress work seamless with it.

@LK4D4 can you take a look at this? I know you've played with this lock in the past.

WeiZhang555 commented:
I've finished testing with `docker-stress` and here is the result:

```
$ ./main -b /usr/bin/docker --config ./stress.json -c 50 --containers 1000 -k 20s
INFO[1010] ran 1053 containers in 1010.11 seconds (1.04 container per sec. or 0.96 sec. per container)
INFO[1010] failures 0
```

docker version result:
```
$ /usr/bin/docker version
Client:
 Version:      1.10.0-dev
 API version:  1.22
 Go version:   go1.5.2
 Git commit:   c11d992
 Built:        Tue Dec 22 02:05:59 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0-dev
 API version:  1.22
 Go version:   go1.5.2
 Git commit:   c11d992
 Built:        Tue Dec 22 02:05:59 2015
 OS/Arch:      linux/amd64
```

commit `c11d992` is branch with my patch exactly:
```
$ git log
commit c11d992de0f66ea9795a626141f29d8298eaabd1
Author: Zhang Wei <zhangwei555@huawei.com>
Date:   Thu Dec 17 11:20:30 2015 +0800

    Break big lock into some tiny locks

    Don't involve code waiting for blocking channel in locked critical
    section because it has potential risk of hanging forever.

    Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

commit 312c82677bdc86d50b483d642ad8c61f1c840c55
Merge: eff810a 843084b
Author: Sebastiaan van Stijn <thaJeztah@users.noreply.github.com>
Date:   Mon Dec 21 23:45:18 2015 +0100

    Merge pull request #15879 from Mashimiao/add-support-blkio_throtte_iops

    Add support for blkio read/write iops device

commit eff810aed688879f67a3730c41d9adce4637470f
```

and content of `./stress.json` is changed a little bit because my machine is locating behind a proxy which has problem visiting `google.com`:
```
[
    {"name":"busybox", "args": ["true"]},
    {"name":"busybox", "args": ["ping", "-c", "5", "10.110.52.38"]},
    {"name":"busybox", "flags": ["--log-driver=syslog"], "args": ["ping", "-c", "5", "10.110.52.38"]}
]
```

Looks like everything works well, I'm happy with the result :laughing: 
WDYT?  @calavera

WeiZhang555 commented:
Just rebased and removed irreleavant changes, `docker stress` still ok:
```
$ ./main -b /usr/bin/docker --config ./stress.json -c 50 --containers 1000 -k 20s
INFO[1008] ran 1053 containers in 1008.04 seconds (1.04 container per sec. or 0.96 sec. per container)
INFO[1008] failures 0
```

WeiZhang555 commented:
Rebased.

thaJeztah commented:
ping @LK4D4 can you have a look at this?

calavera commented:
Thanks @WeiZhang555! Changes LGTM. Let's wait until @LK4D4 can take a look.

WeiZhang555 commented:
ping

thaJeztah commented:
@LK4D4 has some days off until the end of the year, so this probably has to wait a bit. I'll assign him to this PR so that he can have a look when he's back. Thanks for being patient @WeiZhang555

WeiZhang555 commented:
@LK4D4 Have you looked into this? :smile:

thaJeztah commented:
ping @LK4D4 if you have some time :+1:

LK4D4 commented:
LGTM

thaJeztah commented:
counting 2 LGTM's so looks like this is ready to merge :+1:

tonistiigi commented:
@LK4D4 @WeiZhang555 How does this protect us against starting same container twice. Seems to me that lock is released and anyone can call the start method again without the `isRunning` protection.

LK4D4 commented:
@tonistiigi Yeah, seems like we need something like "starting" :)

WeiZhang555 commented:
@LK4D4 Yes, I'm also thinking about this today, I think a new state such as "starting" is necessary, or we can't stop user from starting a container twice.
@tonistiigi Nice catch, thank you!

I'll append a new commit to fix this. :smile:

cpuguy83 commented:
See also #19330
